fix(git config settings pages): Add system-wide settings

mstv opened the pull request:
Fixes #10914
Fixes #11764

## Proposed changes

- Get git settings using `git config list` in order to correctly display effective settings
- Display effective settings read-only (in order to make users explicitly decide in which scope to configure git)
- Display system-wide settings (read-only)
- Add advanced git config "merge.autostash"
- Several refactorings:
  - Extract `IPersistentConfigValueStore` from `IConfigFileSettings`
  - Remove `SettingsSourceExtension.ByPath`
  - Rename to `GitConfigSettingsSet` and to `GitConfigBaseSettingsPage`
  - Add `GitConfig/DistributedSettingsSet.Save`
  - Extract `GitEncodingSettings` from `ConfigFileSettings`

Review commit by commit is recommended.

## Screenshots <!-- Remove this section if PR does not change UI -->

### Before

![image](https://github.com/user-attachments/assets/fd130bb1-d168-4d43-8d2e-910ffbac6cb1)

### After

![image](https://github.com/user-attachments/assets/342b9208-d3db-46c1-baef-6e293d263715)

## Test methodology <!-- How did you ensure quality? -->

- manual

## Please do not squash merge

----

:black_nib: I contribute this code under [The Developer Certificate of Origin](../blob/master/contributors.txt).

pmiossec commented:
> * Display effective settings read-only (in order to make users explicitly decide in which scope to configure git)

😍

mstv commented:
> > * Display effective settings read-only (in order to make users explicitly decide in which scope to configure git)
> 
> 😍

me included

---

Do you have a clue why this PR makes "your" testcase `RebasePatchesBuiltAccordingToGitRebaseFiles` fail, Philippe?

pmiossec commented:
> Do you have a clue why this PR makes "your" testcase `RebasePatchesBuiltAccordingToGitRebaseFiles` fail, Philippe?

The name of the test made me think that it was a flaky test so I relaunched the build.

I didn't remembered I wrote the test and had not looked at the possible reason.
I will need to investigate on a computer (and not from my smartphone)

pmiossec commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -2085,20 +2090,15 @@ public IEnumerable<string> GetSettings(string setting)
         public string? GetGitSetting(string setting, string scopeArg, bool cache = false)
         {
-            GitArgumentBuilder args = new("config") { "--includes", scopeArg, "--get", setting };
+            GitArgumentBuilder args = new("config") { "get", scopeArg, "--includes", setting };
```
```suggestion
            GitArgumentBuilder args = new("config") { "--get", scopeArg, "--includes", setting };
```

pmiossec replied:
Sorry, I was not aware of this git command line change in v2.46
I had to update. So this PR won't work with <v2.46...

mstv replied:
Good to have this caught before merging!

pmiossec commented:
I don't know if the behavior was like that before but now, when switching between git config "setting source", changes are saved immediately (without clicking 'Apply').

mstv commented:
> I don't know if the behavior was like that before but now, when switching between git config "setting source", changes are saved immediately (without clicking 'Apply').

Yes, this behavior has changed.
Global / local settings must be saved in order to update effective settings.
This could be postponed until effective settings become selected. A confirmation popup could be added, too.
But I think this is not needed because other settings are even saved at once.
Thoughts?

RussKie commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -316,11 +317,15 @@ public IConfigFileSettings LocalConfigFile
         public static readonly Encoding LosslessEncoding = Encoding.GetEncoding("ISO-8859-1"); // is any better?
 
-        public Encoding FilesEncoding => ((ConfigFileSettings)EffectiveConfigFile).FilesEncoding ?? new UTF8Encoding(false);
+        private Encoding? _defaultEncoding;
```
[nit] move to the top

RussKie commented on src/app/GitCommands/Git/GitModule.cs:
```diff
@@ -316,11 +317,15 @@ public IConfigFileSettings LocalConfigFile
-        public Encoding FilesEncoding => ((ConfigFileSettings)EffectiveConfigFile).FilesEncoding ?? new UTF8Encoding(false);
+        private Encoding? _defaultEncoding;
+
+        private Encoding DefaultEncoding => _defaultEncoding ??= new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
```
Why is the field needed? Why not init the property straight away? Why not make it static?

mstv replied:
👍
Yes, lazy instantiation is not worth it.

RussKie commented on src/app/GitCommands/Settings/EffectiveGitConfigSettings.cs:
```diff
@@ -0,0 +1,37 @@
+namespace GitCommands.Settings;
+
+[DebuggerDisplay("{" + nameof(DebuggerDisplay) + ",nq}")]
+public sealed class EffectiveGitConfigSettings(IExecutable gitExecutable) : GitConfigSettingsBase(gitExecutable, GitSettingLevel.Effective), IConfigValueStore
```
Does this and the rest of the hierarchy need to be public?
Would be great to have some xml-docs for the types and public members

mstv replied:
`public` because used by GUI project.
I have added xml-docs to classes and to `protected` stuff.
Public methods are interface implementations and inherit the xml-docs.

RussKie commented on src/app/GitCommands/Settings/EffectiveGitConfigSettings.cs:
```diff
@@ -0,0 +1,37 @@
+
+    protected override void Update() => Update(Clear, StoreSetting);
+
+    private void Clear()
```
[nit] Keeping methods in alphabetical order helps navigating the code.

RussKie commented on src/app/GitCommands/Settings/GitConfigSettings.cs:
```diff
@@ -0,0 +1,110 @@
+    {
+        if (gitSettingLevel == GitSettingLevel.Effective)
+        {
+            throw new ArgumentOutOfRangeException(nameof(gitSettingLevel), $"Use read-only {nameof(EffectiveGitConfigSettings)} instance instead.");
```
[`ArgumentOutOfRangeException`](https://learn.microsoft.com/dotnet/api/system.argumentoutofrangeexception?view=net-9.0#methods) has a number of Throw* helpers, may worth using those.

mstv replied:
> [`ArgumentOutOfRangeException`](https://learn.microsoft.com/dotnet/api/system.argumentoutofrangeexception?view=net-9.0#methods) has a number of Throw* helpers, may worth using those.

already checked, not enough: it is unable to compare enum values

RussKie replied:
These aren't flags, so compare as ints?

RussKie commented on src/app/GitUI/CommandsDialogs/SettingsDialog/GitConfigBaseSettingsPage.cs:
```diff
@@ -0,0 +1,69 @@
+using GitExtensions.Extensibility.Settings;
+using Microsoft;
+
+namespace GitUI.CommandsDialogs.SettingsDialog
```
```suggestion
namespace GitUI.CommandsDialogs.SettingsDialog;
```

mstv replied:
Done. I had just renamed this class (in a sub-optimal way, but it is similar to `GitConfigAdvancedSettingsPage`).

RussKie reviewed (commented):
🚀

pmiossec commented on src/app/GitCommands/Git/Commands.Execution.cs:
```diff
@@ -0,0 +1,85 @@
+        bool isSet = !string.IsNullOrEmpty(value);
+        GitArgumentBuilder args = new("config")
+            {
+#if UseNewGitConfigSyntax2_46
```
Is there a particular reason to not use `GitVersion.Current` (i.e. creating a property `GitVersion.Current.SupportNewGitConfigSyntax`) instead?

gerhardol commented:
This does not work with WSL

Some paths like editor and mergetool is expected to be in Windows format.
This change picks up the WSL settings if you edit for a WSL repo, so the Windows settings are garbled.
GE expects Windows paths for these tools.

In some way it is an improvement, but not working yet

gerhardol reviewed (dismissed):
WSL Git settings are messed up, not possible to e.g. merge in WSL

gerhardol commented:
How to proceed with this PR (and the follow up)?
This PR has some issues with WSL (like an issue to save), most is in the follow up.
This could though promote WSL Git from being a semi hidden use case to give full support.

Some issues I see, no full investigation.
* GE uses win-Git also in WSL for some operations. Most prominently for merges. I guess most of this could be implemented in WSL.
* GE reading Git settings (like mergetool path, checking existence) to determine what feature to enable. Ignore for WSL?
* GE saving/updating settings. Only read for WSL?

mstv commented:
I have already changed it to always use the native - i.e. Windows - git for getting & setting git config.

We could add another selection to these two git settings pages which switches between native git and WSL git.

gerhardol commented:
> I have already changed it to always use the native - i.e. Windows - git for getting & setting git config.
> 
> We could add another selection to these two git settings pages which switches between native git and WSL git.

You still get an exception in WSL just switching to local view.
I have not debugged, you have so many changes going on here so it would just be conflicts

GitExtensions.Extensibility.ExternalOperationException
  HResult=0x80131500
  Message=Error getting config values
fatal: --local can only be used inside a git repository


  Source=GitExtensions.Extensibility
  StackTrace:
   at GitExtensions.Extensibility.ExecutionResult.ThrowIfErrorExit(String errorMessage) in C:\dev\gc\gitextensions_4\src\app\GitExtensions.Extensibility\ExecutionResult.cs:line 36
   at GitCommands.Git.Commands.GetGitSettings(IExecutable gitExecutable, GitSettingLevel settingLevel) in C:\dev\gc\gitextensions_4\src\app\GitCommands\Git\Commands.Execution.cs:line 43
   at GitCommands.Settings.GitConfigSettingsBase.Update(Action clear, Action`2 storeSetting) in C:\dev\gc\gitextensions_4\src\app\GitCommands\Settings\GitConfigSettingsBase.cs:line 121
   at GitCommands.Settings.GitConfigSettings.Update() in C:\dev\gc\gitextensions_4\src\app\GitCommands\Settings\GitConfigSettings.cs:line 113
   at GitCommands.Settings.GitConfigSettings.GetValue(String name) in C:\dev\gc\gitextensions_4\src\app\GitCommands\Settings\GitConfigSettings.cs:line 46
   at GitCommands.Settings.SettingsSource`1.GetValue(String name) in C:\dev\gc\gitextensions_4\src\app\GitCommands\Settings\SettingsSource_T.cs:line 17
   at GitCommands.DiffMergeTools.DiffMergeToolConfigurationManager.get_ConfiguredMergeTool() in C:\dev\gc\gitextensions_4\src\app\GitCommands\DiffMergeTools\DiffMergeToolConfigurationManager.cs:line 41
   at GitUI.CommandsDialogs.SettingsDialog.Pages.GitConfigSettingsPage.SettingsToPage() in C:\dev\gc\gitextensions_4\src\app\GitUI\CommandsDialogs\SettingsDialog\Pages\GitConfigSettingsPage.cs:line 84
   at GitUI.CommandsDialogs.SettingsDialog.SettingsPageBase.LoadSettings() in C:\dev\gc\gitextensions_4\src\app\GitUI\CommandsDialogs\SettingsDialog\SettingsPageBase.cs:line 109
   at GitUI.CommandsDialogs.SettingsDialog.GitConfigBaseSettingsPage.SetCurrentSettings(SettingsSource settings) in C:\dev\gc\gitextensions_4\src\app\GitUI\CommandsDialogs\SettingsDialog\GitConfigBaseSettingsPage.cs:line 66
   at GitUI.CommandsDialogs.SettingsDialog.GitConfigBaseSettingsPage.SetLocalSettings() in C:\dev\gc\gitextensions_4\src\app\GitUI\CommandsDialogs\SettingsDialog\GitConfigBaseSettingsPage.cs:line 44
   at GitUI.CommandsDialogs.SettingsDialog.SettingsPageHeader.<>c__DisplayClass5_0.<ConfigureHeader>b__0(Object s, EventArgs e) in C:\dev\gc\gitextensions_4\src\app\GitUI\CommandsDialogs\SettingsDialog\SettingsPageHeader.cs:line 78
   at System.Windows.Forms.RadioButton.OnCheckedChanged(EventArgs e)
   at System.Windows.Forms.RadioButton.set_Checked(Boolean value)
   at System.Windows.Forms.RadioButton.OnClick(EventArgs e)
   at System.Windows.Forms.RadioButton.OnMouseUp(MouseEventArgs mevent)
   at System.Windows.Forms.Control.WmMouseUp(Message& m, MouseButtons button, Int32 clicks)
   at System.Windows.Forms.Control.WndProc(Message& m)
   at System.Windows.Forms.ButtonBase.WndProc(Message& m)
   at System.Windows.Forms.Control.ControlNativeWindow.WndProc(Message& m)
   at System.Windows.Forms.NativeWindow.Callback(HWND hWnd, MessageId msg, WPARAM wparam, LPARAM lparam)

Inner Exception 1:
Exception: Error getting config values
fatal: --local can only be used inside a git repository

mstv commented:
> I have already changed it to always use the native - i.e. Windows - git for getting & setting git config.

This was nonsense because git config settings must be set in WSL. Reverted.
The '$' in "$LOCAL" etc. must be escaped for WSL!
In addition, I have adapted the suggested command lines for diff/merge tools (not fully tested yet.)

> We could add another selection to these two git settings pages which switches between native git and WSL git.

not really necessary:
Git settings pages affect WSL while a WSL repo is loaded.
(Global) Windows git settings can be set using settings pages opened from Dashboard.
(The Windows git cannot be used to access local settings of a WSL repo!)

> GE uses win-Git also in WSL for some operations. Most prominently for merges. I guess most of this could be implemented in WSL.

should be done in a follow-up

> GE reading Git settings (like mergetool path, checking existence) to determine what feature to enable. Ignore for WSL?

We should always use WSL commands. Then only WSL git config settings apply. And these are shown and controlled in the settings pages.

gerhardol commented:
PR seem reasonable (and is an improvement especially for WSL eventually).
However, I cannot take it for a spin yet. Dashboard settings fail (readonly/lazy), WSL fails for me with popup.
(will review after initial tryout).

![image](https://github.com/user-attachments/assets/80cef847-8c98-4609-8afd-b0061fed2fd4)

mstv commented:
> WSL fails for me with popup.

Thank you for testing. This error means that the check for the git version used the wrong one (Windows instead of WSL). And your WSL git seems to be older than 2.46 in contrast to your Windows git.

And yes, (effective and) local settings must not be evaluated if there is no repo. Workaround until fixed: set win-git configs from a non-WSL repo.

gerhardol commented:
> your WSL git seems to be older than 2.46 in contrast to your Windows git.

The version shipped with the distro is what most users will use.
Ubuntu-20.04 with Git 2.25.1 is the oldest I want to support. 
I see Ubuntu 22.04 Git 2.34 as mainline.
At work I have at least 2.46, but may user do not add the separate update sources needed to update
The versions can never be assumed to be the same

>>    GE reading Git settings (like mergetool path, checking existence) to determine what feature to enable. Ignore for WSL?

> We should always use WSL commands. Then only WSL git config settings apply. And these are shown and controlled in the settings pages.

One reason to use Win settings is to keep for instance multiple diff/mergetools common. There has been some users requesting to always use Win configuraion to never configure too (but that would be slower in the general use case).

The special manipulation for merge (and more? do not recall) was not obvious to handle in WSL, but is possible.

gerhardol commented:
Not sure what I want here.
This PR raises WSL Git settings from being hidden, to shown when you start settings in that repo
But the summary is slightly inconsistent now.
difftool/mergetool is reported as different as they are changed.
Git version is still Windows version and ssh uses command line settings only (the env var that controls ssh in Win are not made available in WSL, the check could be removed. (so could a few other checks, but they are not conflicting).

![image](https://github.com/user-attachments/assets/32925797-49e5-4113-a97b-d4c6fb59079c)

Very brief usage, but all seem OK (and minimal review)

mstv commented:
The mount points are different for WSL Ubuntu 24 and 22. Changed to "/mnt/c/..."

gerhardol reviewed (commented):
Finally a re-review (had not posted by re-review and sorry for not removing request changes).

This works well and is an improvement for WSL. Suggest in WSL should be done for Merge too and also directly when selecting tool (I thought it was not working at first)

worktree config is not displayed in the GUI, but handled in effective settings which is certainly good enough

gerhardol commented:
Probably not this PR
Neither is really wrong, should be consistent

![image](https://github.com/user-attachments/assets/3cecd5a0-99cf-442f-ab84-12b6bb9d16c9)

mstv commented:
> Suggest in WSL should be done for Merge too and also directly when selecting tool (I thought it was not working at first)

I had adapted the merge tool suggestion, too. But it did not work because
FormResolveConflicts calls the merge tool directly:

![image](https://github.com/user-attachments/assets/4e9e7856-4289-4f7e-a96a-7dd56431b95c)

BTW: GE must be restarted in order to changed merge tool config to come into effect.
Then:

![image](https://github.com/user-attachments/assets/54d0d27f-da74-4318-a4c6-5142a48f0537)

Is there another way to launch the merge tool in GE?

---

> worktree config is not displayed in the GUI, but handled in effective settings which is certainly good enough

`git config --list --worktree` needs special setup (on Windows):
```
fatal: --worktree cannot be used with multiple working trees unless the config
extension worktreeConfig is enabled. Please read "CONFIGURATION FILE"
section in "git help worktree" for details
```

My 2ct: Worktree config is a niche usecase.

---

> Probably not this PR
> Neither is really wrong, should be consistent

> ![image](https://github.com/user-attachments/assets/3cecd5a0-99cf-442f-ab84-12b6bb9d16c9)

If this was not an addendum to the comments about worktree or merge tool, I would need some more explanation.

mstv commented on src/app/GitCommands/Settings/GitConfigSettingsBase.cs:
```diff
@@ -0,0 +1,131 @@
+        {
+            int firstSeparator = name.IndexOf('.');
+            int lastSeparator = name.LastIndexOf('.');
+            if (lastSeparator < 0 || firstSeparator == lastSeparator)
```
"section" and "setting value" are case-insensitive. Everything in between is the case-sensitive "subsection".

https://git-scm.com/docs/git-config#_syntax

mstv commented on src/app/GitCommands/Settings/EffectiveGitConfigSettings.cs:
```diff
@@ -0,0 +1,42 @@
+
+    private void StoreSetting(string name, string value)
+    {
+        // Entries from all scopes are listed starting with the most generic scope, use the last one, i.e. the most specific
```
It is a comment about the implementation - why it is different from `GitConfigSettings.StoreSetting`.

gerhardol commented:
> > Suggest in WSL should be done for Merge too and also directly when selecting tool (I thought it was not working at first)
> 
> I had adapted the merge tool suggestion, too. But it did not work because FormResolveConflicts calls the merge tool directly:
> 
> Is there another way to launch the merge tool in GE?

Did not recall this.
You can use `Start mergetool`, the second button, then you skip the GE processing of conflicts.
As this executes in Win, the Win command must be used, not the Git (effective) command.
(Same in a few other situations).

Maybe not this PR, just add a comment about it.

> > worktree config is not displayed in the GUI, but handled in effective settings which is certainly good enough

I suggested a comment in a followow-up review

> > Probably not this PR
> > Neither is really wrong, should be consistent

> If this was not an addendum to the comments about worktree or merge tool, I would need some more explanation.

If tool is not set, merge can be modified, diff is readonly

mstv commented:
> If tool is not set, merge can be modified, diff is readonly

not for me (Otherwise, I would have noticed the inconsistency, too, I guess)

![image](https://github.com/user-attachments/assets/6c48dfe6-28d2-4202-8f64-29e649e3079e)

gerhardol commented:
Mergetool to use win need to be handled before next release, if not in this pr
(One alternative is to disable, not my preferred way.)

mstv commented:
> Mergetool to use win need to be handled before next release, if not in this pr

Both ways to start a merge tool work well with this PR (with both suggestions adapting to WSL).
Only the follow-up PR #12201 (branch "git_config") needs something like `GetNativeGitSetting`.

gerhardol commented:
> not for me (Otherwise, I would have noticed the inconsistency, too, I guess)

only in some situations, but now I lost where I got it
another issue...
